drivers/adc/ad9656/ad9656.c

```diff
@@ -0,0 +1,248 @@
+			uint8_t *reg_data)
+{
+	uint8_t buf[3];
+
+	int32_t ret;
+
+	/*
+	 * the MSB of byte 0 indicates a r/w operation, following 7 bits are the
+	 * bits 14-8 of the address of the register that is accessed. Byte 1
+	 * contains the bits 7-0 of the address of the register.
+	 */
+	buf[0] = 0x80 | (reg_addr >> 8);
+	buf[1] = reg_addr & 0xFF;
+	buf[2] = 0x00;
+
+	ret = spi_write_and_read(dev->spi_desc, buf, 3);
+
+	if (ret < 0)
+		return ret;
+
+	*reg_data = buf[2];
+
+	return ret;
+}
+
+/**
+ * @brief Write to the ad9656 that is conected to the SPI
+ * @param dev - The device handler for the ad9656 chip
+ * @param reg_addr - Address of the internal register of the ad9656 chip
+ * @param reg_data - Value to be written to the register
+ * @return SUCCESS if the value was written successfully, FAILURE otherwise
+ */
+int32_t ad9656_reg_write(struct ad9656_dev *dev,
+			 uint16_t reg_addr,
+			 uint8_t reg_data)
+{
+	uint8_t buf[3];
+
+	/*
+	 * the MSB of byte 0 indicates a r/w operation, following 7 bits are the
+	 * bits 14-8 of the address of the register that is accessed. Byte 1
+	 * contains the bits 7-0 of the address of the register. Byte 2 contains
+	 * the data to be written.
+	 */
+	buf[0] = reg_addr >> 8;
+	buf[1] = reg_addr & 0xFF;
+	buf[2] = reg_data;
+
+	return spi_write_and_read(dev->spi_desc, buf, 3);
+}
+
+/**
+ * @brief Launch the JESD204 test functionality for the receive path for the ad9656 chip
+ * @param dev - The device handler for the ad9656 chip
+ * @param test_mode - The type of test that is to be performed or OFF if the testing
+ * 					  process is to be stopped
+ * @return SUCCESS if the ad9656 chip could be successfully set for JESD204 link testing,
+ * 		   FAILURE otherwise
+ */
+int32_t ad9656_JESD204_test(struct ad9656_dev *dev,
+			    uint32_t test_mode)
+{
+	uint8_t format;
+	ad9656_spi_write(dev, AD9656_REG_ADC_TEST_MODE, test_mode);
+	if (test_mode == AD9656_TEST_OFF)
+		format = AD9656_FORMAT_OFFSET_BINARY;
+	else
+		format = AD9656_FORMAT_2S_COMPLEMENT;
+
+	return ad9656_spi_write(dev, AD9656_REG_OUTPUT_MODE, format);
+}
+
+/**
+ * @brief Launch the user input test functionality for the receive path for the ad9656 chip.
+ * 	      User input data is supplied on the outputs instead of real data captured on the ADC.
+ * @param dev - The device handler for the ad9656 chip
+ * @param test_mode - The type of test that is to be performed or OFF if the testing
+ * 					  process is to be stopped
+ * @param user_input_test_pattern - User input test pattern that is sent on the output instead of
+ * 					the ADC	data
+ * @return SUCCESS if the ad9656 chip could be successfully set for user input testing, FAILURE
+ * 		   otherwise
+ */
+int32_t ad9656_user_input_test(struct ad9656_dev *dev, uint32_t test_mode,
+			       struct ad9656_user_input_test_pattern user_input_test_pattern)
+{
+	uint8_t format;
+
+	/* write user input test pattern in the registers */
+	ad9656_spi_write(dev, AD9656_REG_USER_TEST_PATTERN_1_LSB,
```

Comment:
check return for each spi_write. If something fails it make no sense to go further.

Reply:
Done in v6